extmod/vfs_lfsx.c

```diff
@@ -203,8 +217,11 @@ STATIC mp_obj_t MP_VFS_LFSx(ilistdir_func)(size_t n_args, const mp_obj_t *args)
         path = vstr_null_terminated_str(&self->cur_dir);
     }
 
-    MP_VFS_LFSx(ilistdir_it_t) * iter = mp_obj_malloc(MP_VFS_LFSx(ilistdir_it_t), &mp_type_polymorph_iter);
+    MP_VFS_LFSx(ilistdir_it_t) * iter = m_new_obj_with_finaliser(MP_VFS_LFSx(ilistdir_it_t));
+    iter->base.type = &mp_type_polymorph_iter_with_finaliser;
+
     iter->iternext = MP_VFS_LFSx(ilistdir_it_iternext);
+    iter->finaliser = MP_VFS_LFSx(ilistdir_it_del);
     iter->is_str = is_str_type;
     iter->vfs = self;
     int ret = LFSx_API(dir_open)(&self->lfs, &iter->dir, path);
```

Comment:
If `dir_open` fails then `vfs` is still non-NULL and the finaliser will run and call `dir_close`, which may do something bad.

I suggest `iter->vfs` only after this call succeeds.